doc/userguide/firewall/firewall-design.rst

```diff
@@ -0,0 +1,158 @@
+
+The firewall mode in Suricata allows the use of a ruleset that has different
+properties than the default "threat detection" rulesets:
+
+1. default policy is ``drop``, meaning a firewall ruleset needs to specify what
+   is allowed
+2. firewall rules are loaded from separate files
+3. firewall rules use a new action ``accept``
+4. firewall rules are required to use explicit action scopes and rule hooks (see below)
+
+
+
+Actions and Action Scopes
+-------------------------
+
+Firewall rules require action scopes to be explicitly specified.
+
+accept
+~~~~~~
+
+``accept`` is used to issue an accept verdict to the packet, flow or hook.
+
+* ``packet`` accept this packet
+* ``flow`` accept the rest of the packets in this flow
+* ``hook`` accept rules for the current hook/state, evaluate the next hooks
+
+drop
+~~~~
+
+``drop`` is used to drop either the packet or the flow
+
+* ``packet`` drop this packet directly, don't eval any further rules
+* ``flow`` drop this packet as with ``packet`` and drop all future packets in this flow 
+
+.. note:: the action ``pass`` is not available in firewall rules due to ambiguity around
+   the existing meaning for threat detection rules.
+
+
+Explicit rule hook (states)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the regular IDS/IPS rules the engine infers from the rule's matching logic where the
+rule should be "hooked" into the engine. While this works well for these types of rules,
+it does lead to many edge cases that are not acceptable in an firewall ruleset. For this
+reason in the firewall rules the hook needs to be explicitly set.
+
+This is done in the protocol field of the rule. Where in threat detection a rule might look like::
+
+    alert http ... http.uri; ...
+
+In the firewall case it will be::
+
+    accept:hook http1:request_line ... http.uri; ...
+
+The application layer states / hooks are defined per protocol. Each of the hooks has it's own
+default-``drop`` policy, so a ruleset needs a ``accept`` rule for each of the states to allow
+the traffic to flow through.
+
+
+http
+^^^^
+
+For the HTTP protocol there are a number of states to hook into. These apply to HTTP 0.9, 1.0
+and 1.1. HTTP/2 uses it's own state machine.
+
+Available states:
+
+Request (``to_server``) side:
+
+* ``request_not_started``
+* ``request_line``
+* ``request_headers``
+* ``request_body``
+* ``request_trailer``
+* ``request_complete``
+
+Response (``to_client``) side:
+
+* ``response_not_started``
+* ``response_line``
+* ``response_headers``
+* ``response_body``
+* ``response_trailer``
+* ``response_complete``
+
+tls
+^^^
+
+Available states:
+
+Request (``to_server``) side:
+
+* ``client_in_progress``
+* ``client_hello_done``
+* ``client_cert_done``
+* ``client_handshake_done``
+* ``client_finished``
+
+Response (``to_client``) side:
+
+* ``server_in_progress``
+* ``server_hello``
+* ``server_cert_done``
+* ``server_hello_done``
+* ``server_handshake_done``
+* ``server_finished``
+
+Firewall pipeline
```

Comment:
For a future iteration: I miss seeing all possible hooks/ tables in a table or list.